source/concurrent/SynchronizedQueue.ooc

```diff
@@ -67,16 +85,25 @@ BlockedQueue: class <T> extends SynchronizedQueue<T> {
 	}
 	wait: func (isOk := null as Bool*) -> T {
 		result: T = null
-		this _mutex lock()
-		while (this empty) {
+		(e, c) := (this empty, this canceled)
+		while (e && !c) {
```

Comment:
When you think it's empty here, it might actually not be empty since you unlocked the mutex and someone might have added something to the queue immediately after. Then you will sleep and wait to be signaled even though the queue is populated.